client/api/notebook.py

```diff
@@ -53,17 +54,32 @@ def backup(self):
         messages = {}
         self.save_notebook()
         self.assignment.cmd_args.set_args(['--backup'])
-        self.run('file_contents', messages)
+        self.save(messages)
         return self.run('backup', messages)
 
     def submit(self):
         messages = {}
-        self.save_notebook()
         self.assignment.cmd_args.set_args(['--submit'])
-        self.run('file_contents', messages)
+        self.save(messages)
         return self.run('backup', messages)
 
-    def save_notebook(self):
+    def save(self, messages, delay=0.5, retry=3):
+        self.save_notebook(delay)
+        self.run('file_contents', messages)
+        attempts = 0
+        while (attempts <= retry and
+               not validate_contents(messages['file_contents'])):
+            log.info("Notebook file is invalid, Retrying File Read")
+            time.sleep(delay)
+            self.run('file_contents', messages)
+            attempts += 1
+
+        if not validate_contents(messages['file_contents']):
+            log.warning("OK could not autosave the notebook file completely."
+                        " Please ensure that the submission URL on OK appears complete")
+        return messages
```

Comment:
I'm finding this logic hard to follow (since a lot of things are repeated) so here's my suggested version:
```
def save(self, messages, delay=0.5, attempts=3):
    self.save_notebook()
    for _ in range(attempts):
        self.run('file_contents', messages)
        if validate_contents(messages['file_contents'])):
            return messages
        else:
            log.info("Notebook file is invalid, Retrying File Read")
            time.sleep(delay)
    log.warning("OK could not autosave...")
```